Feature/additional components

Tyler-Ward opened the pull request:
Draft implementation of #50.

An exmple file to work with this is shown below.

```yaml
connectors:
  X1:
    type: Powerpole
    pinout: [GND, VCC]
    ignore_in_bom: true
    additional_components:
      -
        type: housing
        qty: 1
        unit: ea
        manufacturer: Anderson
        manufacturer_part_number: 1327G6
        internal_part_number: powerpole-black
      -
        type: housing
        qty: 1
        unit: ea
        manufacturer: Anderson
        manufacturer_part_number: 1327
        internal_part_number: powerpole-red
      -
        type: crimp
        qty: pincount
        unit: ea
        manufacturer: Anderson
        manufacturer_part_number: 1331
        internal_part_number: powerpole-crimp-30A
  X2: &mf3-5
    type: Molex Micro-Fit
    subtype: female
    pinout: [GND, VCC, A, B, C]
    manufacturer_part_number: 436450500
    manufacturer: Molex
    internal_part_number: MF3.0-5
    additional_components:
      -
        type: crimp
        qty: connectioncount
        unit: ea
        manufacturer: Molex
        manufacturer_part_number: 43030-0007
        internal_part_number: MF3.0-Crimp

  X3:
    <<: *mf3-5

cables:
  W1: &wire_power # define template
    colors: [BK, RD] # number of wires implicit in color list
    gauge: 0.25 # assume mm2 if no gauge unit is specified
    show_equiv: true
    length: 0.2
    manufacturer_part_number: 123456

  W2:
    <<: *wire_power

connections:
  -
    - X1: [1-2]
    - W1: [1-2]
    - X2: [1-2]
  -
    - X1: [1-2]
    - W2: [1-2]
    - X3: [1-2]
```

Tyler-Ward commented:
This is now just about functional but needs cleanup of the code, making the output look better, error checking, altered names to match #114 and addition of later sugestions in #50. Will continue this after #114 is done.
![image](https://user-images.githubusercontent.com/3390465/88116590-cd768280-cbb0-11ea-8f38-59de9fedab72.png)

formatc1702 commented:
Not too shabby!

My feeling is that the mini-BOMs within the nodes should be either one line per item (might stretch the nodes, I know),
one multiline cell per item (to easily distinguish separate items visually) or something similar. If I have a new suggestion, I can let you know.

Note: The unit is missing on the cable's sleeving, as well as the crimps and housing (should be showing `ea`)

formatc1702 commented:
Another idea which might keep the diagram simple, but definitely require the BOM table, is the following:

- Auto-assign an Item Number (e.g. 1, 2, 3, ... or 10, 20, 30, ...) to every BOM item, show item number as first column in master BOM
- In the mini-BOMs, show one line = one cell per item in the format:
  `{qty} {unit} x {itemno}` for maximum minimalism
  or
  `{qty} {unit} x {itemno} ({description})` to keep it user-friendly but at least keeping the P/Ns out of it.

One could do a separate cell for each `{}` but hide the vertical borders within a line to get a nice table look without to much visual clutter?

formatc1702 commented:
Quick reminder to use [imperative mood](https://chris.beams.io/posts/git-commit/#imperative) in git commit messages ;)

Tyler-Ward commented:
This branch has now been brought up to date with dev and the rest of the ideas from #50 have been aded in.

Merging the Bom items to a single cell definatley looks better. Using bom indexing to make the graph simpler sounds like a good idea, this will need some refactoring of the BOM generation code to get this to work but should be fairly easy to add once that is done. Other than that I think the only other thing that needs adding is adding a tutorial and syntax description (once #111 is merged in). 

![image](https://user-images.githubusercontent.com/3390465/93692200-031fe500-fae8-11ea-9c06-94bd912e3697.png)

Tyler-Ward commented:
The Bom has been refactored to remove duplications and add an id that can be looekd up when generating a graph. this is then used for the aditional components sections. The aditional component mode can be changed by a harness attribute to slect the option above but this will need #158 or similar  in order to be used

![image](https://user-images.githubusercontent.com/3390465/94341962-ae0f3200-0005-11eb-96ae-ec8ed3e90fac.png)

kvid commented on src/wireviz/Harness.py:
```diff
@@ -394,24 +456,77 @@ def bom(self):
-            if isinstance(item.get('designators', None), List):
-                item['designators'].sort()  # sort designators if a list is provided
             item = {'item': name, 'qty': item.get('qty', None), 'unit': item.get('unit', None), 'designators': item.get('designators', None),
                     'manufacturer': item.get('manufacturer', None), 'mpn': item.get('mpn', None), 'pn': item.get('pn', None)}
```
I suggest removing all 7 `None` arguments in the 4 lines above. The second parameter of `.get()` is optional with default value`=None`. Simpler code is normally easier to read.

kvid reviewed (changes requested):
@Tyler-Ward Thank you for this very useful contribution. I'm just another contributor, like you. @formatc1702 is the owner that approves and merges PRs into dev when he thinks they are ready. He has been on vacation for several weeks, and I decided to review your PR and suggest some changes that I believe will help you getting this PR accepted, but remember that @formatc1702 might not share all my opinions.
- I like the BOM deduplication at the end as it enables removing a lot of other complex deduplication code.
- Consider moving more duplicated code into common (generic if needed) functions.

Tyler-Ward commented:
Thanks for your detailed comments @kvid. They all look good so have integrated them. Moving the components to a dataclass rather than a dict definatly helped the readability of the code in harness.py

kvid commented on src/wireviz/DataClasses.py:
```diff
@@ -7,6 +7,22 @@
+    pn: Optional[str] = None
+    qty: float = 1
+    unit: Optional[str] = None
+    qty_multiplier: Optional[str] = None
```
Consider appending a comment to this line:
```py
# pincount | populated (for connectors), or wirecount | terminations | length | total_length (for cables)
```

kvid reviewed (changes requested):
- Thank you for accepting my suggestions.
- See code comments for some extra simplification suggestion

Tyler-Ward commented:
Thanks for the additional suggestions, Have added them as well, I left the swapping from dict to dataclass code the same as it was for now as the one liner is a bit harder to read but happy to swap to that if other people think that is better.

kvid commented on src/wireviz/Harness.py:
```diff
@@ -97,8 +99,17 @@ def create_graph(self) -> Graph:
+                rows.append(["Additional components"])
+                for extra in connector.additional_components:
+                    qty = extra.qty * connector.get_qty_multiplier(extra.qty_multiplier)
+                    if(self.mini_bom_mode):
```
The usage of parentheses here is very unusual in Python. Maybe you're also a C/C++ programmer as I am? 😄 I suggest removing these parentheses from the `if`-statement.

kvid commented on src/wireviz/Harness.py:
```diff
@@ -97,8 +99,17 @@ def create_graph(self) -> Graph:
+                for extra in connector.additional_components:
+                    qty = extra.qty * connector.get_qty_multiplier(extra.qty_multiplier)
+                    if(self.mini_bom_mode):
+                        id = self.get_bom_index(extra.description, extra.unit, extra.manufacturer, extra.mpn, extra.pn)
+                        rows.append(component_table_entry(f'{id} ({extra.type.capitalize()})', qty, extra.unit))
```
Personally, I would prefer some prefix to the BOM id number in the component table for increased readability, 
e.g. `'#'`, `'No '` or `'BOM Id '`, but other users might disagree (e.g. also US/UK differences), and I suggest supporting an optional string value in the `mini_bom_mode` parameter in addition to the current True and False values. I therefore suggest inserting this code just below the current assignment of the `id` variable:
```py
if isinstance(self.mini_bom_mode, str):
    id = f'{self.mini_bom_mode}{id}'
```
The `if self.mini_bom_mode:` above will still work as intended for both `bool` and `str` values.

kvid commented on src/wireviz/Harness.py:
```diff
@@ -330,90 +351,118 @@ def output(self, filename: (str, Path), view: bool = False, cleanup: bool = True
+        bom_types_group = lambda bt: (bt['item'], bt['unit'], bt['manufacturer'], bt['mpn'], bt['pn'])
+        for group in Counter([bom_types_group(v) for v in bom_items]):
+            items = [v for v in bom_items if bom_types_group(v) == group]
             shared = items[0]
-            designators = [i['designator'] for i in items]
+            designators = []
+            for item in items:
+                if item.get('designators'):
+                    if isinstance(item['designators'], List):
+                        designators.extend(item['designators'])
+                    else:
+                        designators.append(item['designators'])
```
I suggest renaming `items` to `group_entries` and `item` to `inp` to avoid any confusion with the BOM `item` column name.

kvid commented on src/wireviz/Harness.py:
```diff
@@ -330,90 +351,118 @@ def output(self, filename: (str, Path), view: bool = False, cleanup: bool = True
-                    'manufacturer': item.get('manufacturer', None), 'mpn': item.get('mpn', None), 'pn': item.get('pn', None)}
-            bom_extra.append(item)
-        bom_extra = sorted(bom_extra, key=lambda k: k['item'])
-        bom.extend(bom_extra)
-        return bom
+            self._bom.append(item)
```
As above,  I suggest using `entry` as BOM entry variable.

kvid commented on src/wireviz/Harness.py:
```diff
@@ -330,90 +351,118 @@ def output(self, filename: (str, Path), view: bool = False, cleanup: bool = True
+        self._bom = sorted(self._bom, key=lambda k: k['item'])  # sort list of dicts by their values (https://stackoverflow.com/a/73050)
+        # add index
+        index = 1
+        for item in self._bom:
+            item["id"] = index
```
As above,  I suggest using `entry` as BOM entry variable.

kvid commented on src/wireviz/Harness.py:
```diff
@@ -330,90 +351,118 @@ def output(self, filename: (str, Path), view: bool = False, cleanup: bool = True
+        return self._bom
+
+    def get_bom_index(self, item, unit, manufacturer, mpn, pn):
+        for bom_item in self.bom():
+            if((bom_item['item'], bom_item['unit'], bom_item['manufacturer'], bom_item['mpn'], bom_item['pn']) == (item, unit, manufacturer, mpn, pn)):
+                return bom_item['id']
```
As above,  I suggest using `entry` as BOM entry variable.

kvid reviewed (changes requested):
Thank you again for accepting most of my suggestions. I also like that you argue against suggestions that might harm readability. I'm sorry for suggesting more changes so many times, but I often find it hard to see all improvement issues at once. Some issues become much easier too see after some other improvement steps have been done first. I've marked some comments with DRY for [Don't Repeat Yourself](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) because I prefer to avoid repeating the same code twice when it's possible to rewrite it to a loop, function call, or some other way to avoid code duplication without harming readability.

kvid commented on src/wireviz/Harness.py:
```diff
@@ -330,90 +351,118 @@ def output(self, filename: (str, Path), view: bool = False, cleanup: bool = True
+                        'manufacturer': part.manufacturer,
+                        'mpn': part.mpn,
+                        'pn': part.pn,
+                        'designators': cable.name
```
I guess this code line should have ended with `if cable.show_name else None` as for connectors, but by using a common method, this should be fixed.

Tyler-Ward reviewed (commented):
No Problem, thanks for taking the time to keep reviewing. Some more good sugestions here, have implemented most of them and added comments on the others

kvid reviewed (changes requested):
Thank you once again for accepting most of my numerous suggestions. Please let me know when you are tired of improving details in this PR, and prefer raising a new issue to handle further improvements after merge-in, but please take a look at least at the two suggestions beginning with lines 441 and 448 in Harness.py.

_Edit:_ I'm really glad you've managed to both add this useful feature with additional components and that you in the process also simplified the BOM generating code that will be a great benefit to other PRs. The owner wrote in https://github.com/formatc1702/WireViz/issues/101#issuecomment-703774009 that he will finish the remaining PRs of v0.2 first, to avoid "feature creep", and then process the other PRs for v0.3. I suggested to him that your PR is a good candidate to start with, and hope he agrees when he arrives that stage. That means you should be prepared to rebase your PR branch on top of the new `dev` right after the v0.2 release. Thank you for all the good work!

formatc1702 commented:
Progress on this PR looks fantastic, thank you guys for your heavy work so far.
@Tyler-Ward, I hope you're able to include @kvid's latest suggestions soon! 

Fair warning: I might have 1-2 little nit-picks as well before I merge, but @kvid has done the heavy lifting and overall, this looks very good!

Tyler-Ward commented on src/wireviz/Harness.py:
```diff
@@ -330,90 +351,118 @@ def output(self, filename: (str, Path), view: bool = False, cleanup: bool = True
+            item = {'item': item.get('description', ''), 'qty': item.get('qty'), 'unit': item.get('unit'), 'designators': item.get('designators'),
+                    'manufacturer': item.get('manufacturer'), 'mpn': item.get('mpn'), 'pn': item.get('pn')}
+            bom_items.append(item)
+
```
I had missed that the remove_line_breaks function would pass cleanly over non string items. so that would have worked fine.

I have done some tests on fields with newlines. Adding newlines (or any other whitespace changes) in fields will cause bom lines to be split into multiple rows. The catch with multiline fields is that by default the | operator in yaml appends a newline on the end generating additional spaces in the descriptions when run through remove_line_breaks() and causing them not to match due to that. 

There are a few options for how to deal with this.
1. Leave it as is with items being spit /defer this to a new issue.
2. Remove whitespace during deduplication matching, this could cause the unwanted whitespace to appear in the bom descriptions as we would need to pick one to use.
3. move the linebreak code back to the individual fields

I think 3 is probably the best soloution, will implement that tomorow unless there is a preference for one of the options.

Tyler-Ward reviewed (commented):
I have added the majority of the suggestions above, excluding the one that was sigested to skip and the one commented on below which has highlighted some bugs that need fixing which I will sort tomorrow.

> Thank you once again for accepting most of my numerous suggestions. Please let me know when you are tired of improving details in this PR, and prefer raising a new issue to handle further improvements after merge-in, but please take a look at least at the two suggestions beginning with lines 441 and 448 of Harness.py.

Thanks for the additional suggestions, Happy to keep on making edits now if there are further things to improve. or for them to be split into seperate tickets.

> Edit2: One of my code comments in this review is still tagged as Pending which I normally only see on comments before submitting the review, and I wonder if this comment was i fact submitted together with the rest of the review. Can you check if you see my comment as part of this review on line 343 of Harness.py with this contents:

That appears for me and has been fixed (along with a couple of others)

kvid commented on src/wireviz/Harness.py:
```diff
@@ -329,91 +334,126 @@ def output(self, filename: (str, Path), view: bool = False, cleanup: bool = True
+
+    def get_bom_index(self, item, unit, manufacturer, mpn, pn):
+        for entry in self.bom():
+            if(entry['item'], entry['unit'], entry['manufacturer'], entry['mpn'], entry['pn']) == (item, unit, manufacturer, mpn, pn):
```
```suggestion
            if (entry['item'], entry['unit'], entry['manufacturer'], entry['mpn'], entry['pn']) == tuple(clean_whitespace(v) for v in (item, unit, manufacturer, mpn, pn)):
```

kvid commented on src/wireviz/Harness.py:
```diff
@@ -426,5 +466,6 @@ def bom_list(self):
             item_list = [item.get(key, '') for key in keys]  # fill missing values with blanks
             item_list = [', '.join(subitem) if isinstance(subitem, List) else subitem for subitem in item_list]  # convert any lists into comma separated strings
             item_list = ['' if subitem is None else subitem for subitem in item_list]  # if a field is missing for some (but not all) BOM items
+            item_list = [remove_line_breaks(subitem) for subitem in item_list]  # remove line breaks if present
```
If my other suggestion solves your problem, then this can be removed:
```suggestion
```

kvid reviewed (changes requested):
Thank you for accepting most of my suggestions again. Se my suggested solution to the problem you describe. I haven't had time to look at the other modifications you made yet.

Tyler-Ward commented:
That would catch some of the issues, the one I found that is easiest to end up in however is the following where a space is added before a comma. we could add code to replace any space comma combinations with a single comma, not sure if that would cause any other issues but can investigate that later today.

Below is a section of a bom where the connector type was split into two lines

Id | Item | Qty | Unit | Designators | P/N | Manufacturer | MPN
-- | -- | -- | -- | -- | -- | -- | --
3 | Connector, Molex Micro-Fit , female, 5 pins | 2 |   | X2, X3 | MF3.0-5 | Molex | 436450500
4 | Connector, Molex Micro-Fit, female, 5 pins | 1 |   | ZZ | MF3.0-5 | Molex | 436450500

kvid commented:
> That would catch some of the issues, the one I found that is easiest to end up in however is the following where a space is added before a comma. we could add code to replace any space comma combinations with a single comma, not sure if that would cause any other issues but can investigate that later today.

Please share  the YAML input that provoke the undesired output.

> 
> Below is a section of a bom where the connector type was split into two lines
> 
> Id	Item	Qty	Unit	Designators	P/N	Manufacturer	MPN
> 3	Connector, Molex Micro-Fit , female, 5 pins	2	 	X2, X3	MF3.0-5	Molex	436450500
> 4	Connector, Molex Micro-Fit, female, 5 pins	1	 	ZZ	MF3.0-5	Molex	436450500

I assume the first of these two entries was produced by specifying `additional_components` for X2 and X3. Am I right? In that case, the problem is caused by the comma added in `AdditionalComponent.description()`. I suggest calling `clean_whitespace()` on the attribute that is concatenated to the comma separator without a space between, i.e.:
```py
    @property
    def description(self) -> str:
        return clean_whitespace(self.type.capitalize()) + (f', {self.subtype}' if self.subtype else '')
```
I could not find any other code that concatenated a string attribute with a non-space separator before inserting into `Harness.bom_entries`.

Tyler-Ward commented:
The yaml that generated that bom is included at the end of the post (probably not the best test case  in its current form but can be turned into one).

The error is with the bom line for the connector itself (X2 and X3) where the type was split into two lines. ZZ is a aditional bom item that matches the output of a connector without a linebreak in its type. The extra space is added before the comma because the trailing newline is kept when the description string is built from the component variables (line 374 in the current state).

```yaml
connectors:
  X1:
    type: Powerpole
    pinlabels: [GND, VCC]
    ignore_in_bom: true
    additional_components:
      -
        type: housing
        subtype: Anderson Powerpole 15/45, Black
        qty: 1
        manufacturer: Anderson
        mpn: 1327G6
        pn: powerpole-black
      -
        type: housing
        subtype: Anderson Powerpole 15/45, Red
        qty: 1
        manufacturer: Anderson
        mpn: 1327
        pn: powerpole-red
      -
        type: crimp
        subtype: Anderson Powerpole 15/45, 30A
        qty_multiplier: pincount
        manufacturer: Anderson
        mpn: 1331
        pn: powerpole-crimp-30A
    notes: "notes go here"
  X2: &mf3-5
    type: |
      Molex
      Micro-Fit
    subtype: female
    pinlabels: [GND, VCC, A, B, C]
    mpn: 436450500
    manufacturer: Molex
    pn: MF3.0-5
    additional_components:
      -
        type: crimp
        qty_multiplier: populated
        manufacturer: Molex
        mpn: 43030-0007
        pn: MF3.0-Crimp

  X3:
    <<: *mf3-5

  F1:
    style: simple
    autogenerate: true
    type: Crimp ferrule
    subtype: 0.5 mm²
    color: OG # optional color
    additional_components:
      -
        type: crimp housing

cables:
  W1: &wire_power # define template
    colors: [BK, RD] # number of wires implicit in color list
    gauge: 0.25 # assume mm2 if no gauge unit is specified
    show_equiv: true
    length: 0.2
    mpn: 123456

  W2:
    <<: *wire_power
    additional_components:
      -
        type: |
            Sleve
        subtype: |
            Braided nylon,
            black, 3mm
        qty_multiplier: length
        unit: m
        pn: SLV-1

  W3:
    <<: *wire_power

connections:
  -
    - X1: [1-2]
    - W1: [1-2]
    - X2: [1-2]
  -
    - X1: [1-2]
    - W2: [1-2]
    - X3: [1-2]
  -
    - X1: [1-2]
    - W3: [1-2]
    - F1

additional_bom_items:
  - # define an additional item to add to the bill of materials (does not appear in graph)
    description: |
        Connector,
        Molex Micro-Fit, female, 5 pins
    qty: 1
    designators:
      - ZZ
    manufacturer: Molex
    mpn: 436450500
    pn: MF3.0-5
```

kvid commented:
> The yaml that generated that bom is included at the end of the post (probably not the best test case in its current form but can be turned into one).

Thank's for sharing. I believe it's a good candidate for future regression tests - you can suggest an improved version when you have the time.

> 
> The error is with the bom line for the connector itself (X2 and X3) where the type was split into two lines. ZZ is a aditional bom item that matches the output of a connector without a linebreak in its type. The extra space is added before the comma because the trailing newline is kept when the description string is built from the component variables (line 374 in the current state).

You are right. That means we need the same here: Calling `clean_whitespace()` separately on all the attribute that might be concatenated to the comma separator without a space between, i.e. `connector.type`, `connector.subtype`, and `cable.type`.

A better alternative is then to expand `clean_whitespace()` instead of these separate calls (including the separate call in the property method suggested earlier today):
```py
def clean_whitespace(inp):
    return ' '.join(inp.split()).replace(' ,', ',') if isinstance(inp, str) else inp
```
In any case, my other suggestions should still be valid (except the separate call in `AdditionalComponent.description()` if you accept my expanded `clean_whitespace()`).

I assume comma is the only non-space separator that can cause such problems...

Tyler-Ward commented:
I have looked through the fields that support newlines and in the bom they are all folowed by a comma or a space so that whitespace cleanup function will work and has been added.

I have rebased ontop of the latest dev to fix the merge conflict and also added the new syntax additions to the documentation.

kvid commented on src/wireviz/wv_helper.py:
```diff
@@ -152,6 +152,9 @@ def graphviz_line_breaks(inp):
 def remove_line_breaks(inp):
     return inp.replace('\n', ' ').strip() if isinstance(inp, str) else inp
```
I suggest removing the no longer used functions `graphviz_line_breaks()` and `remove_line_breaks()` above.

kvid reviewed (changes requested):
Thank you for accepting some of my suggestions again, and for rebasing on top of the new `dev`.

formatc1702 commented on tutorial/tutorial08.yml:
```diff
@@ -28,6 +36,14 @@ cables:
     manufacturer: [WiresCo,WiresCo,WiresCo,WiresCo] # set a manufacter per wire
```
This is an issue that might have to be considered more generally, perhaps in the context of #56 and #127, where you might want other properties assigned on a per-wire basis.
Just something to keep in mind. I still think @kvid's suggestion could come in handy here.

kvid replied:
My comment above is not about changing any feature. This is a feature that has been implemented several months ago and is present in v0.2. Wire attributes `pn`, `manufacturer`, and `mpn` can all be specified different for each wire in a list, or as a single value common to all wires. Please correct me if I'm wrong. See inserting [common values](https://github.com/formatc1702/WireViz/blob/v0.2/src/wireviz/Harness.py#L163-L166) and [list entries](https://github.com/formatc1702/WireViz/blob/v0.2/src/wireviz/Harness.py#L202-L206) in the diagram, and [inserting either](https://github.com/formatc1702/WireViz/blob/v0.2/src/wireviz/Harness.py#L379-L381) in the BOM.

My suggested comment in the tutorial is just to make the user aware of this feature, but please suggest a different/better way to demonstrate this feature. I'm sorry that the [doc](https://github.com/formatc1702/WireViz/blob/v0.2/docs/syntax.md#cable-attributes) slipped through my review without mentioning this feature.

If adding such a comment is a big issue that needs a discussion, then it's out of scope here. Adding a similar feature for other attributes is definitely out of scope here. 😃

Tyler-Ward replied:
Looking back at the code, there is actualy different functionality in the graph depening on if it is defined as a list or as a sting. If it is a string it will be listed in the top info of a bundle (like it is for a cable), or if it is a list it is listed per wire. The output can be altered so bundles are always rendered per wire even if it is a string, this is probably worth doing as it will avoid bundles looking incosistent.

formatc1702 replied:
@Tyler-Ward has been so kind and updated `syntax.md` to reflect the changes of this PR already, so I would request to add a note regarding this functionality (parameters accepting either `str` or `List[str]`) as well :)

formatc1702 replied:
Sorry, it seems we were typing a comment at the same time.

> The output can be altered so bundles are always rendered per wire even if it is a string, this is probably worth doing as it will avoid bundles looking incosistent.

Let's move that discussion to a new issue/PR so as not to delay this one too much, since it's drifting out of scope a bit. Based on my recent thoughts posted in #56 we'll need to think about the general issue of how to deal with parameters that differ between wires of a bundle, or perhaps even within a regular cable (e.g. different gauges).
So please don't worry about this point right now.

Tyler-Ward replied:
Sounds sensible, will open an issue for this now.

Tyler-Ward commented on src/wireviz/Harness.py:
```diff
@@ -7,10 +7,11 @@
     nested_html_table, flatten2d, index_if_list, html_line_breaks, \
-    graphviz_line_breaks, remove_line_breaks, open_file_read, open_file_write, \
-    html_colorbar, html_image, html_caption, manufacturer_info_field
+    graphviz_line_breaks, remove_line_breaks, clean_whitespace, open_file_read, \
```
I cant find any use of either of those functions in any other locations so is it woerth removing the functions as well?

Tyler-Ward commented on src/wireviz/DataClasses.py:
```diff
@@ -43,6 +46,21 @@ def __post_init__(self, gv_dir):
+
+    @property
+    def description(self) -> str:
+        return self.type.capitalize().strip() + (f', {self.subtype.strip()}' if self.subtype else '')
```
This is needed to avoid some graphicial odities in the diagram. Without the strip if either string is a multiline string (defined with | as in the examples) a traling newline is added which causes issues when the fields are joined together in the cell. Most of the other cells ignore this problem as the trailing <br/> is ignored by graphvis.

with mini bom
![image](https://user-images.githubusercontent.com/3390465/96635111-5bc6e580-1313-11eb-8174-08f382c971bb.png)
without mini bom
![image](https://user-images.githubusercontent.com/3390465/96635252-85800c80-1313-11eb-9a89-469ba03ef24e.png)

The alternatives could be either getting user to use the |- symbol instead (should update the docs in this case). or moving the cleanup of the "<br />,", "<br />)" and "<br /><br />" into the render code for the additional components cells.

formatc1702 replied:
In this case I'd argue having that little `.rstrip()` is nicer than requiring `|-` in the input YAML.

formatc1702 replied:
My question is, where does the need for `.capitalize()` come from?
Perhaps I'm missing something, but I don't see a need to mess with the user-defined `type` attribute. This applies to all instances of this function call.

Tyler-Ward replied:
The capitialise was originaly added to create more consistent bom formatting. We can remove this and pass that requirment onto the user if they care about it as this would be fairly intuative for a user to know what to change (unlike the newlines).

kvid replied:
I agree with your decision to remove the `.capitalize()`. Personally, I think I would have preferred recommending `|-` in the doc before removing the trailing newline with `.rstrip()`, but I understand that others might have a different opinion. The only motivation for a user to use `|` when specifying this attribute is to insert newlines, and then it shouldn't be too much of a surprise seeing the results above. If we in the doc write "recommend `|-` to avoid a trailing newline", that should be enough, IMHO, but I respect that you disagree. No hard feelings. 😄

formatc1702 commented on src/wireviz/DataClasses.py:
```diff
@@ -7,6 +7,9 @@
 from wireviz.wv_helper import int2tuple, aspect_ratio
 from wireviz import wv_colors
 
+ConnectorMultiplier = str  # = Literal['pincount', 'populated']
+CableMultiplier = str  # = Literal['wirecount', 'terminations', 'length', 'total_length']
+
```
Any reason the `Literal`s are commented out?

Tyler-Ward replied:
from @kvid's comment https://github.com/formatc1702/WireViz/pull/115#discussion_r497943873
> Note that the Literal type specifier is in comments to avoid requiring Python 3.8.

I have checked my 2.7 install and literals aernt supported, doing some searching I cant find any reference to this method of specifying literals so not sure if this is a standard implementation method or not.

formatc1702 replied:
Understood. Let's keep it like this then, perhaps add a comment behind the comment explaining this.

kvid replied:
New in version 3.8 is specified by https://docs.python.org/3/library/typing.html#typing.Literal

Tyler-Ward commented:
Thanks for the additional suggestions, I believe all of them have now been implemented or decided against in the inline comments but let me know if I have missed any.

formatc1702 commented:
Once there are no further comments, I will likely squash this PR into a single commit, unless @Tyler-Ward wants to go through the trouble (not worth it IMHO) of compressing the current 30 commits into something more manageable with meaningful intermediate steps.
@kvid since you've been keeping an eye on the implementation, I will rely on your GO to proceed with the squash+merge.

kvid reviewed (approved):
Thank you for accepting more suggestions. As far as I can see now, you have either implemented or decided against all my suggestions. I provided a link to the documentation of `Literal` that you couldn't find in that code comment thread, and explained why I would prefer recommending `|-` before calling `.rstrip()`, but that I also respect your opinion.

Thank you for the fantastic effort to develop this PR with the useful feature and improving the BOM code in the process. I'm also impressed by your patience to process my numerous suggestions about details.
